DeepSpeech.py

```diff
@@ -694,594 +651,999 @@ def stopwatch(start_duration=0):
+        # Find first running epoch that provides a next job
+        for epoch in self._epochs_running:
+            job = epoch.get_job(worker)
+            if job:
+                return job
+        # No next job found
+        return None
+
+    def get_job(self, worker=0):
+        '''Retrieves the first job for a worker.
+
+        Kwargs:
+            worker (int): index of the worker to get the first job for
+
+        Returns:
+            WorkerJob. a job of one of the running epochs that will get
+                associated with the given worker and put into state 'running'
+        '''
+        # Let's ensure that this does not interfer with other workers/requests
+        with self._lock:
+            if is_chief:
+                # First try to get a next job
+                job = self._get_job(worker)
+                if job is None:
+                    # If there was no next job, we give it a second chance by triggering the epoch state machine
+                    if self._next_epoch():
+                        # Epoch state machine got a new epoch
+                        # Second try to get a next job
+                        job = self._get_job(worker)
+                        if job is None:
+                            # Albeit the epoch state machine got a new epoch, the epoch had no new job for us
+                            log_error('Unexpected case - no job for worker %d.' % (worker))
+                        return job
+                    else:
+                        # Epoch state machine has no new epoch
+                        # This happens at the end of the whole training - nothing to worry about
+                        log_traffic('No jobs left for worker %d.' % (worker))
+                        self._log_all_jobs()
+                        return None
+                else:
+                    # We got a new jub from one of the currently running epochs
+                    log_traffic('Got new %s' % str(job))
+                    return job
+            else:
+                # We are a remote worker and have to hand over to the chief worker by HTTP
+                result = self._talk_to_chief(PREFIX_GET_JOB + str(FLAGS.task_index))
+                if result:
+                    result = pickle.loads(result)
+                return result
+
+    def next_job(self, job):
+        '''Sends a finished job back to the coordinator and retrieves in exchange the next one.
+
+        Kwargs:
+            job (WorkerJob): job that was finished by a worker and who's results are to be
+                digested by the coordinator
+
+        Returns:
+            WorkerJob. next job of one of the running epochs that will get
+                associated with the worker from the finished job and put into state 'running'
+        '''
+        if is_chief:
+            # Try to find the epoch the job belongs to
+            epoch = next((epoch for epoch in self._epochs_running if epoch.id == job.epoch_id), None)
```

Comment:
Same comment as for the `jobs_open/running/done` lists. Can we make them `dict`s?